Helpers Best Practices Guide (#12556)
SatelliteQE/robottelo: BEST_PRACTICES.md

## Introduction
The scattered and misaligned placement of helpers within the framework can lead to challenges in discovering and managing helpers, resulting in duplicated definitions and increased maintenance overhead. To address these issues, we have outlined best practices for defining helper functions, modules, classes and fixtures. These guidelines will help you make informed decisions about where to place different types of helpers within the framework.

## General Practices
Here are some general best practices for helper management:
- **Prefer API Calls**: When creating test setups, prioritize using API calls (nailgun methods) over UI and CLI interactions, as they tend to be faster and more efficient, unless there is a specific need for specific endpoint based setups. UI setups should be the last resort.

> - **Utilize Target SAT Fixtures**: Ensure that all tests make use of target satellite fixtures to access a wide range of helper methods provided by the satellite object.

Reviewer: let's use the in-code name when mentioning specific fixtures.
```suggestion
- **Utilize `target_sat` Fixtures**: Ensure that all tests make use of target satellite fixtures to access a wide range of helper methods provided by the satellite object.
```

- **Non-Reusable helpers**: If helper/fixture is not reusable, keep them closer to test in test module itself.

> - **No one liner functions**: If the helper function is just one-liner, It's recommended to directly replace the function code where its used.

Reviewer: ```suggestion
- **No one liner functions**: If the helper function is just a one-liner, It's recommended to directly replace the function code where it's used.
```

  - Helpers added to this class or inherited by this class can be accessed as `sat_obj.api_factory.helper_func_name()`.

- `cli_factory.py`
  - This module contains the `CLIFactory` class. Helpers within this class enable integrated operations between components of the satellite object using satellite CLI hammer commands.
  - While it is not our primary choice to add more CLI-based helpers, this class is the appropriate place to add them if the test demands it.
  - Helpers added to this class or inherited by this class can be accessed as `sat_obj.cli_factory.helper_func_name()`.

- `ui_factory.py`
  - This module contains a `UIFactory` class that accepts a session object as a parameter. Helpers within this class enable integrated operations between components of the satellite object using the satellite UI.
  - Although adding more UI-based helpers is not our preference, this class is suitable for adding them if the test demands it.
  - Helpers added to this class or inherited by this class can be accessed as `sat_obj.ui_factory(session).helper_func_name()`.

>   - The session object provided here is the existing session from the test case where the helper is being accessed.

Reviewer: based on our current conversations, it would likely be good to add a line about passing a different session object into the helper when needed.

- **Capsule**:
  - This class represents the Capsule host and includes essential properties and methods that describe the Capsule host.
  - Properties like `url` and `is_stream` specify details about the specific Capsule host.
  - The `satellite` property returns the satellite object to which the Capsule is connected.
  - Methods within this class manage operations related to the Capsule, especially for reconfiguring it via installation processes. Other operational methods are part of the Capsule mixins. e.g `sat.restart_services()`
  - This class inherits methods from ContentHost class, hence methods from these classes are available with this class objects.
- **ContentHost**:
  - This class represents the ContentHost or Client, including crucial properties and methods that describe these hosts.
  - Properties like `hostname` specify details about specific ContentHosts.
  - The `nailgun_host` property returns the API/nailgun object for the ContentHost.
  - Methods within this class manage operations on ContentHosts, particularly for reconfiguring them. Other operational methods are part of the ContentHost mixins.

>   - This class inherits methods from `Brokers` modules `Host` class, hence methods from that class are available with this class objects.

Reviewer: I think it would be good to reverse the order of these and state that each higher level inherits the behavior of the previous host class.

By adhering to these best practices, you'll contribute to the efficient organization and accessibility of helpers within the framework, enhancing the overall development experience and maintainability
 of the project. Thank you for your commitment to best practices!

## FAQs:

> _**Q1. When should I prefer/not prefer fixtures over to helper function when implementing a new helper ?**_

Reviewer: I don't think we need to number the questions. This would also keep us consistent with your formatting for answers.
```suggestion
_**Question: When should I prefer/not prefer fixtures over to helper function when implementing a new helper ?**_
```

_Answer:_ The answers could be multiple. Fixture over functions are preferred when:
   - A new helper function should be cached based on scope of the function usage.
   - A new helper has dependency on other fixtures.
   - A new helper provides the facility of setup and teardown in the helper function itself.
   - A new helper as a fixture accept parameters, allowing you to create dynamic test scenarios.

_**Q2. Where should I implement the fixture that could be used by all tests from all endpoints?**_

>   _Answer:_ The global package `pytest_plugins` is the recommended place. Choose the right subpackage `core` if framework fixture `component` if component fixture.

Reviewer: ```suggestion
  _Answer:_ The global package `pytest_fixtures` is the recommended place. Choose the right sub-package `core` if framework fixture `component` if component fixture.
```

_**Q3. Where should we implement the helper functions needed by fixtures ?**_

>   _Answer:_ It is prefered to use utils but if the function is not reusable one can keep in the fixture module and access locally.

Reviewer: I'd suggest that, if the helper function is only usable by the fixture, then it should live in the fixture's own module. However, if it could be used elsewhere in robottelo/tests then it should become a util.

_**Q4. Whats the most preferred user interface for writing helper that could be used across all three interfaces tests ?**_

  _Answer:_ API helpers are preferred but if test demands setup from UI/CLI interfaces then the helper should be added in those interfaces.

> _**Q5. I want to create a property / method to describe satellite/capsul/host, where should I add that property ?**_

Reviewer: ```suggestion
_**Q5. I want to create a property / method to describe satellite/capsule/host, where should I add that property ?**_
```

  _Answer:_ Satellite/Caspule/ContentHost classes (but not in their mixins) are the best to add properties/methods that describe those.

_**Q6. I want to perform operations on Satellite/Capsule/ContentHost like execute CLI commands, read conf files etc, where should I add a method for that?**_

>   _Answer:_ Add into existing mixin classes in mixin modules in `robottelo/host_helpers/*_mixins.py` or create a new targetted mixin class and inherit than in host classes of `robottelo/hosts.py` module

Reviewer: ```suggestion
  _Answer:_ Add into existing mixin classes in mixin modules in `robottelo/host_helpers/*_mixins.py` or create a new targeted mixin class and inherit than in host classes of `robottelo/hosts.py` module
```

_**Q8. When should I prefer helper function as a host object method over utils function ?**_

  _Answer:_ When the new function is dependent on existing host methods or attributes then its good implement such helper function in host classes as it makes to access those host methods and attributes easier.

_**Q9. What helper functions goes into `robottelo/utils/__init__.py` module?**_
  _Answer:_ If the helper does not fit into any of the existing utils modules or subject of helper(s) is not big enough to create new , then add in `robottelo/utils/__init__.py` as a standalone helper.

_**Q10. I have a helper function that applies to all Satellite, Capsule, ContentHost classes. Which class should I choose to add it?**_

  _Answer:_ The recommeded class in such cases is always the highest parent class. Here ContentHost class is the highest parent class where this function should be added as a class method or a mixin method.

_**Q11. I have some upgrade scenario helpers to implement, where should I add them?**_

  _Answer:_ Upgrade scenarios are not special in this case. All helpers/function except framework should be same as being used in foreman tests and hence all rules are applied.

> _**Q12. Where does a non-reusable helper function should reside?**_

Reviewer: ```suggestion
_**Q12. Where should a non-reusable helper function reside?**_
```

  _Answer:_ The preffered place is the test module where its being used.

_**Q13. I need to extend the functionality of third party library methods/objects, should I do it in robottelo?**_

  _Answer:_ Its recommended to extend the functionality of third library methods in that library itself if its being maintained by SatelliteQE like airgun,nailgun, manifester or is active community like widgetastic, wrapanapi. Else extend that appropriately in utils, host methods or fixtures.

_**Q14. What if I see two helper methods/functions almost doing equal operations with a little diff ?**_

>   _Answer:_ See if this could be merged in one function with optional parameters else leave them separated. E.g provisioning functions using API calls with minimum difference and being used in API, UI and CLI tests then merge them. But if two provisioing helpers one for CLI and API tests and tests demands it then good to keep it separate.

Reviewer: ```suggestion
  _Answer:_ See if this could be merged in one function with optional parameters else leave them separated. E.g provisioning functions using API calls with minimum difference and being used in API, UI and CLI tests then merge them. But if two provisioning helpers one for CLI and API tests and tests demands it then good to keep it separate.
```
